### PR TITLE
feature: raise error for unknown properties in job config

### LINKED ISSUE #303: Resource objects such as *JobConfig do not error or warn when setting unknown properties
Thanks for stopping by to let us know something could be better!

**PLEASE READ**: If you have a support contract with Google, please create an issue in the [support console](https://cloud.google.com/support/) instead of filing on GitHub. This will ensure a timely response.

Please run down the following list and make sure you've tried the usual "quick fixes":

  - Search the issues already opened: https://github.com/googleapis/python-bigquery/issues
  - Search StackOverflow: https://stackoverflow.com/questions/tagged/google-cloud-platform+python

If you are still having issues, please be sure to include as much information as possible:

#### Environment details

  - OS type and version: MACOS Catalina 10.15.5
  - Python version: 3.7.6
  - pip version: 20.0.2
  - `google-cloud-bigquery` version: 1.28.0

#### Steps to reproduce

Step 1.  Extract BQ tabke data to bucket

#### Code example
```
# from google.cloud import bigquery
# client = bigquery.Client()
# bucket_name = 'my-bucket'

destination_uri = "gs://{}/{}".format(bucket_name, "shakespeare.json")
dataset_ref = bigquery.DatasetReference(project, dataset_id)
table_ref = dataset_ref.table("shakespeare")
job_config = bigquery.job.ExtractJobConfig()
job_config.destinationFormat = bigquery.DestinationFormat.NEWLINE_DELIMITED_JSON
job_config.compression = bigquery.Compression.GZIP

extract_job = client.extract_table(
    table_ref,
    destination_uri,
    # Location must match that of the source table.
    location="US",
    job_config=job_config,
)  # API request
extract_job.result()  # Waits for job to complete.

```

  2. Fetch the table from bucket and load into bq

#### Code example

```
from google.cloud import bigquery

# Construct a BigQuery client object.
client = bigquery.Client()

# TODO(developer): Set table_id to the ID of the table to create.
# table_id = "your-project.your_dataset.your_table_name"

job_config = bigquery.LoadJobConfig(
    schema=[
        bigquery.SchemaField("name", "STRING"),
        bigquery.SchemaField("post_abbr", "STRING"),
    ],  #This is not shakespeare schema, as just copied pasted the code provided on google docs
    source_format=bigquery.SourceFormat.NEWLINE_DELIMITED_JSON,
)
uri = "gs://{}/{}".format(bucket_name, "shakespeare.json")

load_job = client.load_table_from_uri(
    uri,
    table_id,
    location="US",  # Must match the destination dataset location.
    job_config=job_config,
)  # Make an API request.

load_job.result()  # Waits for the job to complete.

destination_table = client.get_table(table_id)
print("Loaded {} rows.".format(destination_table.num_rows))


```

#### Stack trace
```
Error while reading data, error message: JSON table encountered too many errors, giving up. Rows: 1; errors: 1. Please look into the errors[] collection for more details.

Error while reading data, error message: JSON processing encountered too many errors, giving up. Rows: 1; errors: 1; max bad: 0; error percent: 0

Error while reading data, error message: JSON parsing error in row starting at position 0: Parser terminated before end of string
```

However when I extract the table using bq command line and follow the example of:

```
bq extract \
--compression GZIP \
--destination_format NEWLINE_DELIMITED_JSON \
'mydataset.mytable' \
gs://example-bucket/myfile.json

``` 

And then run the second step, I don't get any error and my table is successfully imported.

Making sure to follow these steps will guarantee the quickest resolution possible.

Thanks!

### PR DESCRIPTION
**Pull request by cguardia**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Simple feature to raise warning if unknown property is set on job_config instances.

Fixes #303 🦕

### CODE REVIEW

File: google/cloud/bigquery/job/base.py [hunk 1]
```diff
@@ -659,6 +660,12 @@ def __init__(self, job_type, **kwargs):
         for prop, val in kwargs.items():
             setattr(self, prop, val)
 
+    def __setattr__(self, name, value):
+        """Override to be able to warn if an uknown property is being set"""
+        if not name.startswith("_") and name not in type(self).__dict__:
+            warnings.warn("Property {} is unknown for {}.".format(name, type(self)))
```
Comment (tswast) [inline google/cloud/bigquery/job/base.py:666]: On second thought, I think we can go a step further and throw `AttributeError`, just like `object()` instances do.

Comment (cguardia) [reply google/cloud/bigquery/job/base.py]: OK. Done.

**Review by tswast — APPROVED** [review]
Thanks!

**Comment by tswast** [comment]
```
    def __setattr__(self, name, value):
        """Override to be able to raise error if an unknown property is being set"""
        if not name.startswith("_") and name not in type(self).__dict__:
>           raise AttributeError("Property {} is unknown for {}.".format(name, type(self)))
E           AttributeError: Property labels is unknown for <class 'google.cloud.bigquery.job.query.QueryJobConfig'>.
```

**Comment by tswast** [comment]
It appears you've uncovered a bug in the client. I don't see a setter for labels!

https://github.com/googleapis/python-bigquery/blob/0337ea0bde966c3ccb94960493a6fa6f2bee49b4/google/cloud/bigquery/job/base.py#L234

**Comment by tswast** [comment]
Nevermind, there is a setter: https://github.com/googleapis/python-bigquery/blob/0337ea0bde966c3ccb94960493a6fa6f2bee49b4/google/cloud/bigquery/job/base.py#L675

Perhaps this logic needs to be adjusted in order to account for properties defined in the superclass?

File: google/cloud/bigquery/job/base.py [hunk 2]
```diff
@@ -659,6 +659,12 @@ def __init__(self, job_type, **kwargs):
         for prop, val in kwargs.items():
             setattr(self, prop, val)
 
+    def __setattr__(self, name, value):
+        """Override to be able to raise error if an unknown property is being set"""
+        if not name.startswith("_") and name not in type(self).__dict__:
```
Comment (tswast) [inline google/cloud/bigquery/job/base.py:664]: Let's try using [hasattr](https://docs.python.org/3/library/functions.html#hasattr) to see if that accounts for superclass properties.

```suggestion
        if not name.startswith("_") and not hasattr(self, name):
```

**Review by tswast — CHANGES_REQUESTED** [review]

File: google/cloud/bigquery/job/base.py (hunk 2 above)
Comment (cguardia) [reply google/cloud/bigquery/job/base.py]: Ok, but we have to use type(self), because if not the properties are looked up on the instance.

**Review by tswast — APPROVED** [review]

**Comment by tswast** [comment]
One of our system tests actually had a mistake, which this change caught. There is no "destination" property on LoadJobConfig https://googleapis.dev/python/bigquery/latest/generated/google.cloud.bigquery.job.LoadJobConfig.html

**Comment by plamut** [comment]
Re: labels error - it's actually a [bug](https://github.com/googleapis/python-bigquery/issues/469) in the base `_AsyncJob`.

**Pull request merged**